File: .planning/measurements/bonsai-parity/qmm_shapes_bench.py
"""Isolated quantized_matmul kernel bench for Bonsai-Q1 decode shapes.

Mirror of crates/higgs-models/tests/qmm_shapes_bench.rs. Same shapes,
same protocol (warmup 50, then 1000 iters with mx.eval per call).

Run:
    source ~/Dev/diffusion_bonsai/.venv/bin/activate
    python .planning/measurements/bonsai-parity/qmm_shapes_bench.py
"""
from __future__ import annotations

import time

import mlx.core as mx

SHAPES = [
    # Bonsai-1.7B
    ("1p7b/q_or_o",  2048, 2048),
    ("1p7b/k_or_v",  2048, 1024),
    ("1p7b/gate_up", 2048, 6144),
    ("1p7b/down",    6144, 2048),
    ("1p7b/lm_head", 2048, 151669),
    # Bonsai-8B
    ("8b/q_or_o",    4096, 4096),
    ("8b/k_or_v",    4096, 1024),
    ("8b/gate_up",   4096, 12288),
    ("8b/down",      12288, 4096),
    ("8b/lm_head",   4096, 151669),
]

GROUP_SIZE = 128
BITS = 1
WARMUP = 50
ITERS = 1000


def main() -> None:
    print()
    print(f"qmm-isolated bench (Python mlx-py {mx.__version__}) "
          f"group={GROUP_SIZE} bits={BITS}")
    print(f"warmup={WARMUP} iters={ITERS} (default device)")
    print()
    print(f"{'shape':<18} {'K':>6} {'M':>6} {'ms/iter (sync)':>14}")

    for label, K, M in SHAPES:
        w_full = mx.random.normal(shape=(M, K)).astype(mx.float16)
        qw, scales, biases = mx.quantize(w_full, group_size=GROUP_SIZE, bits=BITS)
        mx.eval(qw, scales, biases)

        x = mx.random.normal(shape=(1, 1, K)).astype(mx.float16)
        mx.eval(x)

        for _ in range(WARMUP):
            y = mx.quantized_matmul(
                x, qw, scales=scales, biases=biases,
                transpose=True, group_size=GROUP_SIZE, bits=BITS,
            )
            mx.eval(y)

        t0 = time.perf_counter()
        for _ in range(ITERS):
            y = mx.quantized_matmul(
                x, qw, scales=scales, biases=biases,
                transpose=True, group_size=GROUP_SIZE, bits=BITS,
            )
            mx.eval(y)
        ms_sync = (time.perf_counter() - t0) * 1e3 / ITERS

        print(f"{label:<18} {K:>6} {M:>6} {ms_sync:>14.4f}")


if __name__ == "__main__":
    main()

File: crates/higgs-models/tests/qmm_shapes_bench.rs
//! Isolated `quantized_matmul` kernel bench for Bonsai-Q1 decode shapes.
//!
//! Run:
//!   cargo test --release -p higgs-models --test qmm_shapes_bench \
//!       qmm_shapes_isolated -- --ignored --nocapture
//!
//! Why: end-to-end Bonsai-8B decode is 3.2x slower than mlx-lm Python on the
//! same kernels. This bench isolates qmm cost per shape so we can compare
//! apples-to-apples against `qmm_shapes_bench.py` (mlx 0.31.2.dev). If the
//! Rust numbers match Python at the same shape, the gap is in composition
//! (cache update, rope offset Array allocs, FFI shape queries). If Rust is
//! slower per call, the gap is in mlx-sys / MLX 0.31.1 vs 0.31.2.dev.

use mlx_rs::{ops, random, transforms::eval, Array, Dtype};

#[derive(Clone, Copy)]
struct Shape {
    label: &'static str,
    k: i32,
    m: i32,
}

const SHAPES: &[Shape] = &[
    // Bonsai-1.7B (hidden=2048, intermediate=6144, kv_dim=1024, vocab=151669)
    Shape { label: "1p7b/q_or_o",  k: 2048, m: 2048 },
    Shape { label: "1p7b/k_or_v",  k: 2048, m: 1024 },
    Shape { label: "1p7b/gate_up", k: 2048, m: 6144 },
    Shape { label: "1p7b/down",    k: 6144, m: 2048 },
    Shape { label: "1p7b/lm_head", k: 2048, m: 151669 },
    // Bonsai-8B (hidden=4096, intermediate=12288, kv_dim=1024, vocab=151669)
    Shape { label: "8b/q_or_o",    k: 4096,  m: 4096 },
    Shape { label: "8b/k_or_v",    k: 4096,  m: 1024 },
    Shape { label: "8b/gate_up",   k: 4096,  m: 12288 },
    Shape { label: "8b/down",      k: 12288, m: 4096 },
    Shape { label: "8b/lm_head",   k: 4096,  m: 151669 },
];

const GROUP_SIZE: i32 = 128;
const BITS: i32 = 1;
const WARMUP: usize = 50;
const ITERS: usize = 1000;

fn fresh_x(k: i32) -> Array {
    let x = random::normal::<f32>(&[1i32, 1, k], None, None, None).unwrap();
    x.as_dtype(Dtype::Float16).unwrap()
}

#[test]
#[ignore]
fn qmm_shapes_isolated() {
    println!();
    println!("qmm-isolated bench — group={GROUP_SIZE} bits={BITS}");
    println!("warmup={WARMUP} iters={ITERS} (release build, default device)");
    println!();
    println!(
        "{:<18} {:>6} {:>6} {:>14}",
        "shape", "K", "M", "ms/iter (sync)"
    );

    for shape in SHAPES {
        let w_full = random::normal::<f32>(&[shape.m, shape.k], None, None, None).unwrap();
        let w_full = w_full.as_dtype(Dtype::Float16).unwrap();
        let (qw, scales, biases) = ops::quantize(&w_full, GROUP_SIZE, BITS).unwrap();
        eval([&qw, &scales, &biases]).unwrap();

        let x = fresh_x(shape.k);
        eval([&x]).unwrap();

        for _ in 0..WARMUP {
            let y = ops::quantized_matmul(
                &x, &qw, &scales, Some(&biases), true, GROUP_SIZE, BITS,
            )
            .unwrap();
            eval([&y]).unwrap();
        }

        let t0 = std::time::Instant::now();
        for _ in 0..ITERS {
            let y = ops::quantized_matmul(
                &x, &qw, &scales, Some(&biases), true, GROUP_SIZE, BITS,
            )
            .unwrap();
            eval([&y]).unwrap();
        }
        let ms_sync = t0.elapsed().as_secs_f64() * 1e3 / ITERS as f64;

        println!(
            "{:<18} {:>6} {:>6} {:>14.4}",
            shape.label, shape.k, shape.m, ms_sync
        );
    }
}
